torch/csrc/autograd/profiler.cpp

```diff
@@ -274,6 +288,66 @@ struct ProfilerThreadLocalState
   }
 
  private:
+  void setSourceLocation(Event& evt, const at::RecordFunction& fn) {
+    std::lock_guard<std::mutex> guard(state_mutex_);
+    if (fn.scope() == at::RecordScope::BACKWARD_FUNCTION) {
+      // if that's a backward function then take the source location
+      // from the corresponding function
+      auto fwd_pair = std::make_pair(fn.seqNr(), fn.forwardThreadId());
+      auto it = fwd_source_locs_.find(fwd_pair);
+      if (it != fwd_source_locs_.end()) {
+        evt.setSourceLocation(it->second);
+      }
+      // in some cases we won't find the corr. forward node (e.g. GraphRoot)
+    } else {
+      auto ts_loc = torchScriptCallSite();
+      auto py_stack = pythonCallstack();
+
+      evt.setSourceLocation(EventSourceLocation{ts_loc, py_stack});
+      // if there's a sequence number, save the location for the backward node
+      if (fn.seqNr() >= 0) {
+        fwd_source_locs_[std::make_pair(fn.seqNr(), fn.threadId())] = evt.sourceLocation();
+      }
+    }
+  }
+
+  std::string torchScriptCallSite() {
+    std::stringstream loc;
+    auto call_info = jit::currentCallSite();
+    if (call_info) {
+      auto& range = call_info->source_range;
+      if (range.source()) {
+        auto& src = range.source();
+        if (src && src->filename()) {
+          auto line = src->starting_line_no() +
+              src->lineno_for_offset(range.start());
+
+          loc << *(src->filename()) << "(" << line << "): " << call_info->function_name;
+        }
+      }
+    }
+    return loc.str();
+  }
+
+  std::vector<std::string> pythonCallstack() {
+    std::vector<std::string> py_stack;
+
+    PyGILState_STATE gil_state = PyGILState_Ensure();
```

Comment:
have we timed how slow is it to decode stuff to strings? it might be worthwhile to just bump the refcount on the `f_code` objects and keep them around and decode in post-processing. That's what lsprof seems to do: https://github.com/python/cpython/blob/master/Modules/_lsprof.c#L108 and the processing in https://github.com/python/cpython/blob/master/Lib/cProfile.py#L122

keeping the pointer to the entire PyFrameObject is probably not cool as it might keep all locals vars around

cc @zdevito as a resident cpython expert

Reply:
also you could try to write it with pybind - might be a bit more readable, up to you